### PR TITLE
Cluster topology deletion prevention

### PR DESCRIPTION
**Pull request by shayasoolin**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Implements the ClusterTopology controller with deletion prevention logic. The controller:
- Automatically manages the `grove.io/clustertopology` finalizer on all ClusterTopology resources
- Prevents deletion when the topology is in use by PodCliqueSet resources (via `grove.io/topology-name` label)
- Prevents deletion when the topology is actively configured in the operator settings
- Requeues deletion attempts every 30s until blockers are cleared

Follows existing Grove controller patterns and achieves 90.6% test coverage.

#### Special notes for your reviewer:

**Dependencies:**
- Requires the TopologyConfiguration API changes (separate PR)

**Deletion Prevention Logic:**
The controller uses a conservative approach - it blocks deletion if any PodCliqueSet references the topology OR if the topology is configured in operator settings. The default topology name is `grove-topology` if not specified.

**Test Coverage:**
25 test cases covering creation, deletion, edge cases, and error handling. Uses fake clients following existing Grove test patterns.

#### Additional documentation e.g., enhancement proposals, usage docs, etc.:

```docs
To reference a ClusterTopology from a PodCliqueSet, add the grove.io/topology-name label.

To delete a ClusterTopology, ensure no PodCliqueSet resources reference it and it's not configured in the operator settings. The controller will automatically retry deletion every 30 seconds.
```

### CODE REVIEW

File: operator/internal/controller/clustertopology/reconciledelete.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	ctrlutils "github.com/ai-dynamo/grove/operator/internal/controller/utils"
+
+	"github.com/go-logr/logr"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// triggerDeletionFlow handles the deletion of a ClusterTopology with deletion prevention checks.
+func (r *Reconciler) triggerDeletionFlow(ctx context.Context, logger logr.Logger, ct *grovecorev1alpha1.ClusterTopology) ctrlcommon.ReconcileStepResult {
+	deleteStepFns := []ctrlcommon.ReconcileStepFn[grovecorev1alpha1.ClusterTopology]{
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete.go:37]: Consider adding a log message at the start of deletion flow for better observability.

File: operator/internal/controller/clustertopology/reconciledelete.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	if len(pcsList.Items) > 0 {
+		logger.Info("Cannot delete ClusterTopology: referenced by PodCliqueSet resources",
+			"topologyName", ct.Name,
+			"podCliqueSetCount", len(pcsList.Items))
+		return ctrlcommon.ReconcileAfter(30*time.Second, "topology referenced by PodCliqueSet resources")
+	}
+
+	// Condition 2: Check if topology is enabled and configured to use this ClusterTopology
+	if r.config.Enabled {
+		configuredName := "grove-topology" // default
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete.go:74]: Should be a constant.

File: operator/internal/controller/clustertopology/reconciledelete_test.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+				testutils.NewPodCliqueSetBuilder("test-pcs-3", "other-ns", uuid.NewUUID()).
+					WithTopologyLabel("test-topology").
+					Build(),
+			},
+			wantRequeue: true,
+			wantError:   false,
+		},
+		{
+			name:                "blocked by topology enabled with default name",
+			clusterTopologyName: "grove-topology",
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete_test.go:92]: Default constant?

File: operator/internal/controller/clustertopology/reconciledelete_test.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+				Enabled: true,
+				Name:    ptr.To("custom-topology"),
+			},
+			existingPodCliqueSets: []client.Object{},
+			wantRequeue:           true,
+			wantError:             false,
+		},
+		{
+			name:                "allowed when topology enabled but different name",
+			clusterTopologyName: "grove-topology",
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete_test.go:114]: Default constant?

File: operator/internal/controller/clustertopology/reconciledelete_test.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+					WithTopologyLabel("test-topology").
+					Build(),
+			},
+			wantFinalizerRemoved: false,
+			wantRequeue:          true,
+			wantError:            false,
+		},
+		{
+			name: "blocked deletion flow - topology enabled",
+			ct: testutils.NewClusterTopologyBuilder("grove-topology").
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete_test.go:284]: default constant?

File: operator/internal/controller/clustertopology/reconciledelete.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+func (r *Reconciler) checkDeletionConditions(ctx context.Context, logger logr.Logger, ct *grovecorev1alpha1.ClusterTopology) ctrlcommon.ReconcileStepResult {
+	// Condition 1: Check for PodCliqueSet references
+	pcsList := &grovecorev1alpha1.PodCliqueSetList{}
+	labelSelector := client.MatchingLabels{
+		apicommon.LabelTopologyName: ct.Name,
+	}
+	if err := r.client.List(ctx, pcsList, labelSelector); err != nil {
+		return ctrlcommon.ReconcileWithErrors("failed to list PodCliqueSet resources", err)
+	}
+	if len(pcsList.Items) > 0 {
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete.go:65]: Can we emit an event for operators/admins here?

File: operator/internal/controller/clustertopology/reconciledelete.go [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	// Condition 2: Check if topology is enabled and configured to use this ClusterTopology
+	if r.config.Enabled {
+		configuredName := "grove-topology" // default
+		if r.config.Name != nil {
+			configuredName = *r.config.Name
+		}
+		if configuredName == ct.Name {
+			logger.Info("Cannot delete ClusterTopology: topology feature is enabled and configured to use this topology",
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete.go:79]: Emit an event there as well.

File: operator/internal/controller/clustertopology/reconciledelete.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+				"topologyName", ct.Name,
+				"configuredTopologyName", configuredName)
+			return ctrlcommon.ReconcileAfter(30*time.Second, "topology feature configured to use this ClusterTopology")
+		}
+	}
+
+	logger.Info("ClusterTopology can be safely deleted", "topologyName", ct.Name)
+	return ctrlcommon.ContinueReconcile()
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete.go:87]: Hmm, is there a chance of race between PCS being created and the ClusterTopology being removed? 

1. Reconciler checks and finds 0 PodCliqueSets
2. User creates a PodCliqueSet with this topology
3. Reconciler removes finalizer
4. ClusterTopology is deleted
5. PodCliqueSet now has an invalid topology reference

Just incase,  update the PCS validating webhook to prevent referencing a CT that is being deleted but not finalized yet?

File: operator/internal/controller/clustertopology/reconciler.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+	ct := &grovecorev1alpha1.ClusterTopology{}
+	if err := r.client.Get(ctx, req.NamespacedName, ct); err != nil {
+		return ctrl.Result{}, ctrlclient.IgnoreNotFound(err)
+	}
+
+	// Handle deletion
+	if !ct.DeletionTimestamp.IsZero() {
+		if !controllerutil.ContainsFinalizer(ct, constants.FinalizerClusterTopology) {
+			return ctrl.Result{}, nil
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciler.go:61]: Add a log debug only message for the early exit case to aid debugging?
```suggestion
             	logger.V(1).Info("ClusterTopology marked for deletion but finalizer already removed", "name", ct.Name)
			return ctrl.Result{}, nil
```

File: operator/internal/controller/clustertopology/reconciledelete_test.go [hunk 10]
```diff
@@ -0,0 +1,338 @@
...
+			},
+			existingPodCliqueSets: []client.Object{
+				testutils.NewPodCliqueSetBuilder("test-pcs", "test-ns", uuid.NewUUID()).
+					WithTopologyLabel("different-topology").
+					Build(),
+			},
+			wantRequeue: false,
+			wantError:   false,
+		},
+	}
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete_test.go:147]: Consider adding a test case where the topology is enabled and the name is explicitly set to `ptr.To("grove-topology")` (not nil) to ensure both code paths (nil default vs explicit value) work identically.

File: operator/test/utils/clustertopology.go [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+	}
+
+	return ct
+}
+
+// NewSimpleClusterTopology creates a basic ClusterTopology for testing.
+func NewSimpleClusterTopology(name string) *grovecorev1alpha1.ClusterTopology {
+	return NewClusterTopologyBuilder(name).
+		WithLevels([]grovecorev1alpha1.TopologyLevel{
+			{
```
Comment (gflarity) [inline operator/test/utils/clustertopology.go:84]: The same topology levels are defined in three helper functions (`NewSimpleClusterTopology`, `NewClusterTopologyWithFinalizer`, `NewClusterTopologyMarkedForDeletion`). Consider extracting this to a helper function:

```go
func defaultTopologyLevels() []grovecorev1alpha1.TopologyLevel {
	return []grovecorev1alpha1.TopologyLevel{
		{
			Domain: grovecorev1alpha1.TopologyDomainZone,
			Key:    "topology.kubernetes.io/zone",
		},
		{
			Domain: grovecorev1alpha1.TopologyDomainHost,
			Key:    "kubernetes.io/hostname",
		},
	}
}
```

File: operator/internal/controller/clustertopology/reconciledelete.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+		}
+	}
+
+	logger.Info("ClusterTopology deleted successfully")
+	return ctrlcommon.DoNotRequeue()
+}
+
+// checkDeletionConditions verifies that ClusterTopology can be safely deleted.
+// Deletion is blocked if:
+// 1. Any PodCliqueSet references this topology (via grove.io/topology-name label)
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete.go:54]: Nit, move 1. and 2. right above the code below so the more descriptive explanation is closer to the code in question. Then there's no need for duplicating.

File: operator/internal/controller/clustertopology/reconciledelete.go [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+
+// removeFinalizer removes the ClusterTopology finalizer if present.
+func (r *Reconciler) removeFinalizer(ctx context.Context, logger logr.Logger, ct *grovecorev1alpha1.ClusterTopology) ctrlcommon.ReconcileStepResult {
+	if !controllerutil.ContainsFinalizer(ct, constants.FinalizerClusterTopology) {
+		logger.Info("Finalizer not found", "ClusterTopology", ct.Name)
+		return ctrlcommon.ContinueReconcile()
+	}
+	logger.Info("Removing finalizer", "ClusterTopology", ct.Name, "finalizerName", constants.FinalizerClusterTopology)
+	if err := ctrlutils.RemoveAndPatchFinalizer(ctx, r.client, ct, constants.FinalizerClusterTopology); err != nil {
+		return ctrlcommon.ReconcileWithErrors("error removing finalizer", fmt.Errorf("failed to remove finalizer: %s from ClusterTopology: %s: %w", constants.FinalizerClusterTopology, ct.Name, err))
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconciledelete.go:98]: Logging and events for these errors?

File: operator/internal/controller/clustertopology/reconcilespec.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+	log.Info("Finished spec reconciliation flow", "ClusterTopology", ct.Name)
+	return ctrlcommon.ContinueReconcile()
+}
+
+// ensureFinalizer adds the ClusterTopology finalizer if not already present.
+func (r *Reconciler) ensureFinalizer(ctx context.Context, logger logr.Logger, ct *grovecorev1alpha1.ClusterTopology) ctrlcommon.ReconcileStepResult {
+	if !controllerutil.ContainsFinalizer(ct, constants.FinalizerClusterTopology) {
+		logger.Info("Adding finalizer", "finalizerName", constants.FinalizerClusterTopology)
+		if err := ctrlutils.AddAndPatchFinalizer(ctx, r.client, ct, constants.FinalizerClusterTopology); err != nil {
+			return ctrlcommon.ReconcileWithErrors("error adding finalizer", fmt.Errorf("failed to add finalizer: %s to ClusterTopology: %s: %w", constants.FinalizerClusterTopology, ct.Name, err))
```
Comment (gflarity) [inline operator/internal/controller/clustertopology/reconcilespec.go:55]: Seems like there should be logging for this happening, and perhaps an event?

File: operator/internal/controller/clustertopology/reconciledelete.go (hunk 2 above)
Comment (shayasoolin) [reply operator/internal/controller/clustertopology/reconciledelete.go:74]: Definitely, I am waiting for a different PR to be merged with this constants as well as some other structs I rely on, to rebase and update the code.

File: operator/internal/controller/clustertopology/reconciledelete_test.go (hunk 3 above)
Comment (shayasoolin) [reply operator/internal/controller/clustertopology/reconciledelete_test.go:92]: Will be provided by another PR (not mine).

File: operator/internal/controller/clustertopology/reconciledelete.go (hunk 8 above)
Comment (shayasoolin) [reply operator/internal/controller/clustertopology/reconciledelete.go:87]: Amazing @gflarity, love it. @shmuel-runai this should be implemented in your current PR please.

**Comment by renormalize** [comment]
Superseded by #259.
